### PR TITLE
Remove using an iterator from updating TWAP records

### PR DESCRIPTION
**Pull request by ValarDragon**
This shows up in benchmarks as being a slowdown on getting TWAP records. It will affect gas, but should only do so in EndBlock logic which should not be gas metered.

We see this iterator call being the cause of around .4% of State machine processing speed and even persists in the IAVL v2 benchmarks as a real expense.

### CODE REVIEW

File: x/twap/store.go [hunk 1]
```diff
@@ -151,6 +151,23 @@ func (k Keeper) GetAllMostRecentRecordsForPool(ctx sdk.Context, poolId uint64) (
 	return types.GetAllMostRecentTwapsForPool(store, poolId)
 }
 
+// GetAllMostRecentRecordsForPool returns all most recent twap records
+// (in state representation) for the provided pool id.
+func (k Keeper) GetAllMostRecentRecordsForPoolWithDenoms(ctx sdk.Context, poolId uint64, denoms []string) ([]types.TwapRecord, error) {
```
Comment (mattverse) [inline x/twap/store.go:156]: Can we add test case for this method?

File: x/twap/store.go [hunk 2]
```diff
@@ -151,6 +151,23 @@ func (k Keeper) GetAllMostRecentRecordsForPool(ctx sdk.Context, poolId uint64) (
...
+	store := ctx.KVStore(k.storeKey)
+	// if length != 2, use iterator
```
Comment (mattverse) [inline x/twap/store.go:158]: nice

**Comment by czarcas7ic** [comment]
I think we should add changelog for this, but is soft suggestion

**Review by czarcas7ic — APPROVED** [review]
LGTM, I think we should have the unit test as Matt suggested though.

File: x/twap/types/keys.go [hunk 3]
```diff
@@ -92,6 +92,12 @@ func GetAllMostRecentTwapsForPool(store sdk.KVStore, poolId uint64) ([]TwapRecor
 	return osmoutils.GatherValuesFromStore(store, []byte(startPrefix), []byte(endPrefix), ParseTwapFromBz)
 }
 
+func GetMostRecentTwapForPool(store sdk.KVStore, poolId uint64, denom1, denom2 string) (TwapRecord, error) {
+	key := FormatMostRecentTWAPKey(poolId, denom1, denom2)
+	bz := store.Get(key)
+	return ParseTwapFromBz(bz)
+}
```
Comment (p0mvn) [inline x/twap/types/keys.go:99]: nit: would it be better to have this within keeper since it touches store?

**Comment by faddat** [comment]
Agree about the unit test

File: x/twap/types/keys.go (hunk 3 above)
Comment (mattverse) [reply x/twap/types/keys.go:99]: I think having this here makes sense given prev code structure (e.g GetAllMostRecentTwapsForPool also lives in types/keys.go)

**Comment by mattverse** [comment]
Added unit tests, will be merging this PR by EOW

**Review by czarcas7ic — APPROVED** [review]

**Comment by ValarDragon** [comment]
Thank you for adding the test :pray:

**Pull request merged**